character-controller.js

```diff
@@ -1266,92 +1367,99 @@ class RemotePlayer extends InterpolatedPlayer {
+        const voiceSpec = e.changes.keys.get('voiceSpec');
+        const json = JSON.parse(voiceSpec.value);
+        if (json.endpointUrl)
+          this.loadVoiceEndpoint(json.endpointUrl);
+        if (json.audioUrl && json.indexUrl)
+          this.loadVoicePack({ audioUrl: json.audioUrl, indexUrl: json.indexUrl });
+      }
+
+      if (e.changes.keys.get('name')) {
+        this.name = e.changes.keys.get('name');
+      }
+
+      if (e.changes.keys.has("transform")) {
+        const transform = this.playerMap.get("transform");
+        if (transform) {
+          const remoteTimeDiff = transform[7];
+          
+          this.position.fromArray(transform, 0);
+
+          if (this.avatar) this.characterPhysics.setPosition(this.position);
+
+          this.quaternion.fromArray(transform, 3);
+
+          this.positionInterpolant?.snapshot(remoteTimeDiff);
+          this.quaternionInterpolant?.snapshot(remoteTimeDiff);
+
+          for (const actionBinaryInterpolant of this
+            .actionBinaryInterpolantsArray) {
+            actionBinaryInterpolant.snapshot(remoteTimeDiff);
+          }
+
+          if (this.avatar) {
+            this.avatar.setVelocity(
+              remoteTimeDiff / 1000,
+              lastPosition,
+              this.position,
+              this.quaternion
+            );
+          }
+
+          this.characterPhysics.applyAvatarPhysicsDetail(true, true, performance.now(), remoteTimeDiff / 1000);
+
+          lastPosition.copy(this.position);
+
+        } else {
+          console.warn("Unhandled event", e)
+        }
+      }
     };
-    this.playerMap.observe(observePlayerFn);
-    this.unbindFns.push(this.playerMap.unobserve.bind(this.playerMap, observePlayerFn));
-    
-    this.appManager.bindState(this.getAppsState());
-    this.appManager.loadApps();
-    
-    this.syncAvatar();
-  }
-}
-/* class StaticUninterpolatedPlayer extends UninterpolatedPlayer {
-  constructor(opts) {
-    super(opts);
 
-    this.actions = [];
-  }
-  getActionsState() {
-    return this.actions;
-  }
-  getActions() {
-    return this.actions;
-  }
-  getActionsArray() {
-    return this.actions;
-  }
-  getAction(type) {
-    return this.actions.find(action => action.type === type);
-  }
-  getActionByActionId(actionId) {
-    return this.actions.find(action => action.actionId === actionId);
-  }
-  hasAction(type) {
-    return this.actions.some(a => a.type === type);
-  }
-  addAction(action) {
-    this.actions.push(action);
+    this.playerMap.observe(observePlayerFn);
+    this.unbindFns.push(
+      this.playerMap.unobserve.bind(this.playerMap, observePlayerFn)
+    );
 
-    this.dispatchEvent({
-      type: 'actionadd',
-      action,
-    });
-  }
-  removeAction(type) {
-    for (let i = 0; i < this.actions.length; i++) {
-      const action = this.actions[i];
-      if (action.type === type) {
-        this.removeActionIndex(i);
-        break;
+    this.appManager.callBackFn = (app, event, flag) => {

```

Comment:
If you wanted events you could have a real `EventTarget` but even that seems wrong. I don't know what this is supposed to do but it can be written in a better way.